### PR TITLE
Remove Python 3.6 support

### PR DESCRIPTION
**Pull request by rwgk**
## Description
Remove Python 3.6 support entirely. ([Official end-of-life](https://devguide.python.org/versions/): 2021-12-23)

NOTE: We're losing testing with GCC 4.8 (`centos:7`), but we're deliberately deferring reviewing the C++ code to remove related `#ifdef`s.
________

Stage 1: Change the Python version guard, then fix up GitHub Actions.

```C++
#if PY_VERSION_HEX < 0x03070000
#    error "PYTHON < 3.7 IS UNSUPPORTED. pybind11 v2.12 was the last to support Python 3.6."
#endif
```

First round end of stage 1: @ commit 51851f43af9c5117cb0788c71e038921c8b52fa6
________

Stage 2: Review & fix-as-needed the matching lines:

```bash
#! /bin/bash
set -x
git grep WITH_THREAD
git grep PY_VERSION_HEX
git grep PY_MINOR_VERSION
git grep PYPY_VERSION
git grep -I -E '3\.6'
git grep -I -E '3\.7'
git grep -I -E '\(3, 6'
git grep -I -E '\(3, 7'
git grep -I -E '3[^A-Za-z0-9.]+6'
git grep -I -E '3[^A-Za-z0-9.]+7'
```
________

## Suggested changelog entry:

```rst
Support for Python 3.6 was removed. (Official end-of-life: 2021-12-23)
```

### CODE REVIEW

**Comment by rwgk** [comment]
@henryiii Apparently commit 247496954e437343ae857673022f421d819c1d11 (changing pyproject.toml) didn't change the outcomes of the tests in any way. I'm not sure what that means.

**Comment by henryiii** [comment]
https://github.com/pybind/pybind11/blob/186df220fd06381944ded2f2e65f346d209d648a/setup.cfg#L17

https://github.com/pybind/pybind11/blob/186df220fd06381944ded2f2e65f346d209d648a/setup.cfg#L42

This is what needs changing, forgot we still use setup.cfg/setup.py.

File: tests/test_numpy_array.py [hunk 1]
```diff
@@ -198,7 +198,6 @@ def assert_references(a, b, base=None):
         assert a.flags.f_contiguous == b.flags.f_contiguous
         assert a.flags.writeable == b.flags.writeable
         assert a.flags.aligned == b.flags.aligned
-        # 1.13 supported Python 3.6
         if tuple(int(x) for x in np.__version__.split(".")[:2]) >= (1, 14):
             assert a.flags.writebackifcopy == b.flags.writebackifcopy
```
Comment (henryiii) [inline tests/test_numpy_array.py:202]:
```suggestion
        assert a.flags.writebackifcopy == b.flags.writebackifcopy
```

The oldest numpy to support 3.7 was 1.14 (https://github.com/scipy/oldest-supported-numpy/blob/23ece5a161d5957f5142991393d2ae689ee45b9d/setup.cfg#L56C12-L56C18)

**Review by henryiii — APPROVED** [review]

**Comment by rwgk** [comment]
Thanks for the fixes! Please merge if you see that all GHA pass before I do. (I'll be out for an hour or so.)

**Comment by henryiii** [comment]
I might not be up that much longer. :) We need to add explicit declaration of 3.13 support in the classifiers before a release.

**Pull request merged**